src/polymer-project.ts

```diff
@@ -26,25 +31,186 @@ const extensionsForType = {
+  shell: string;
+  fragments: string[];
+  sourceGlobs: string[];
+  dependencyGlobs: string[];
 
   _splitFiles: Map<string, SplitFile> = new Map();
   _parts: Map<string, SplitFile> = new Map();
-  split = new Splitter(this);
-  rejoin = new Rejoiner(this);
+  _analyzer: StreamAnalyzer;
+  _bundler: Bundler;
+
+  constructor(options?: ProjectOptions) {
+    this.root = options.root || process.cwd();
+    this.entrypoint = osPath.resolve(this.root, options.entrypoint);
+    this.shell = osPath.resolve(this.root, options.shell);
+    this.fragments = (options.fragments || [])
+        .map((f) => osPath.resolve(this.root, f));
+    this.sourceGlobs = (options.sourceGlobs || defaultSourceGlobs)
+        .map((g) => resolveGlob(this.root, g));
+    this.dependencyGlobs = (options.dependencyGlobs || defaultDependencyGlobs)
+        .map((g) => resolveGlob(this.root, g));
+
+    this._analyzer = new StreamAnalyzer(
+      this.root,
+      this.entrypoint,
+      this.shell,
+      this.fragments);
+
+    this._bundler = new Bundler(
+      this.root,
+      this.entrypoint,
+      this.shell,
+      this.fragments,
+      this._analyzer);
+
+    logger.debug(`root: ${this.root}`);
+    logger.debug(`shell: ${this.shell}`);
+    logger.debug(`entrypoint: ${this.entrypoint}`);
+    logger.debug(`fragments: ${this.entrypoint}`);
+    logger.debug(`sources: ${this.sourceGlobs}`);
+    // logger.debug(`dependencies: \n\t${this.dependencyGlobs}`);
+  }
+
+  getSourceGlobs(): string[] {
+    let sourceGlobs = [];
+    if (this.entrypoint) sourceGlobs.push(this.entrypoint);
+    if (this.shell) sourceGlobs.push(this.shell);
+    if (this.fragments && this.fragments.length) {
+      sourceGlobs = sourceGlobs.concat(this.fragments);
+    }
+    if (this.sourceGlobs && this.sourceGlobs.length> 0) {
+      sourceGlobs = sourceGlobs.concat(this.sourceGlobs);
+    }
+    if (this.dependencyGlobs && this.dependencyGlobs.length > 0) {
+      let excludes = this.dependencyGlobs.map((g) => invertGlob(g));
+      logger.debug(`excludes: \n\t${excludes.join('\n\t')}`);
+      sourceGlobs = sourceGlobs.concat(excludes);
+    }
+    logger.debug(`sourceGlobs: \n\t${sourceGlobs.join('\n\t')}`);
+    return sourceGlobs;
+  }
+
+  // TODO(justinfagnani): add options, pass to vfs.src()
+  sources() {
+    return vfs.src(this.getSourceGlobs(), {
+      allowEmpty: true,
+      cwdbase: true,
+      nodir: true,
+    });
+  }
+
+  getDependencyGlobs(): string[] {
+    logger.debug(`dependencyGlobs: \n\t${this.dependencyGlobs.join('\n\t')}`);
+    return this.dependencyGlobs;
+  }
+
+  // TODO(justinfagnani): add options, pass to vfs.src()
+  dependencies(): NodeJS.ReadableStream {
+    let deps = this.getDependencyGlobs();
+    return vfs.src(deps, {
+      allowEmpty: true,
+      cwdbase: true,
+      nodir: true,
+    });
+  }
+
+  // TODO(justinfagnani): add allFiles()
+
+  split(): Transform {
+    return new Splitter(this);
+  }
+
+  rejoin(): Transform {
```

Comment:
I tried to make them all nullary methods, but for the analyzer and bundler they needed to be specific instances because they reference each other. Splitter and joiner on the other hand need to be separate instances because they appear in multiple forks of the build pipeline.